src/ILVerify/src/ILImporter.StackValue.cs

```diff
@@ -92,17 +105,17 @@ static public StackValue CreateFromType(TypeDesc type)
                 case TypeFlags.UInt16:
                 case TypeFlags.Int32:
                 case TypeFlags.UInt32:
-                    return CreatePrimitive(StackValueKind.Int32);
+                    return CreatePrimitive(StackValueKind.Int32, type);
```

Comment:
It does not sound right that you have started tracking the exact type here. The verifier should be based on the verification types only.